source/crud/bulk-write.md

```diff
@@ -0,0 +1,724 @@
+     * lightweight handle to a client with a custom write concern configured, e.g. a
+     * MongoClient.withWriteConcern() method.
+     */
+    writeConcern: Optional<WriteConcern>;
+
+    /**
+     * Whether detailed results for each successful operation should be included in the returned
+     * BulkWriteResult.
+     *
+     * Defaults to false.
+     */
+    verboseResults: Optional<Boolean>;
+}
+```
+
+### Result
+
+```typescript
+class BulkWriteResult {
+    /**
+     * Indicates whether this write result was acknowledged. If not, then all other members of this
+     * result will be undefined.
+     *
+     * NOT REQUIRED TO IMPLEMENT. See the CRUD specification for more guidance on modeling
+     * unacknowledged results.
+     */
+    acknowledged: Boolean;
+
+    /**
+     * Indicates whether the results are verbose. If false, the insertResults, updateResults, and
+     * deleteResults fields in this result will be undefined.
+     *
+     * NOT REQUIRED TO IMPLEMENT. See below for other ways to differentiate summary results from
+     * verbose results.
+     */
+    hasVerboseResults: Boolean;
+
+    /**
+     * The total number of documents inserted across all insert operations.
+     */
+    insertedCount: Int64;
+
+    /**
+     * The total number of documents upserted across all update operations.
+     */
+    upsertedCount: Int64;
+
+    /**
+     * The total number of documents matched across all update operations.
+     */
+    matchedCount: Int64;
+
+    /**
+     * The total number of documents modified across all update operations.
+     */
+    modifiedCount: Int64;
+
+    /**
+     * The total number of documents deleted across all delete operations.
+     */
+    deletedCount: Int64;
+
+    /**
+     * The results of each individual insert operation that was successfully performed.
+     */
+    insertResults: Map<Int64, InsertOneResult>;
+
+    /**
+     * The results of each individual update operation that was successfully performed.
+     */
+    updateResult: Map<Int64, UpdateResult>;
+
+    /**
+     * The results of each individual delete operation that was successfully performed.
+     */
+    deleteResults: Map<Int64, DeleteResult>;
+}
+
+class InsertOneResult {
+    /**
+     * The _id of the inserted document.
+     */
+    insertedId: Any;
+}
+
+class UpdateResult {
+    /**
+     * The number of documents that matched the filter.
+     */
+    matchedCount: Int64;
+
+    /**
+     * The number of documents that were modified.
+     */
+    modifiedCount: Int64;
+
+    /**
+     * The number of documents that were upserted.
+     *
+     * NOT REQUIRED TO IMPLEMENT. Drivers may choose not to provide this property so long as it is
+     * always possible to discern whether an upsert took place.
+     */
+    upsertedCount: Int64;
+
+    /**
+     * The _id field of the upserted document if an upsert occurred.
+     */
+    upsertedId: Optional<Any>;
```

Comment:
> what is the API impact? Are you referring to the spec document itself?

I prefer leaving `upsertedCount` out of the spec unless needed. I expect `upsertedCount` can only be 0 or 1. `upsertedCount` is motivated by supporting upserting an `_id` with BSON NULL (seems rare IMO).

I interpret "NOT REQUIRED TO IMPLEMENT" to mean drivers are free to implement regardless of whether it is needed.

> This would require an envelope object in PHP, which would be clunkier than using a native `null` value and providing an `upsertedCount` field

Given the need, here is a revised suggestion:

```typescript
/**
 * The _id field of the upserted document if an upsert occurred.
 * It MUST be possible for a caller to discern whether an upsert took place, even if the `upsertedId` is a BSON NULL value.
 * If necessary, drivers MAY add a boolean `didUpsert` field.
*/
upsertedId: Optional<Any>;
```

Reply:
Fine to remove the explicit `upsertedCount` in exchange for some extra text under `upsertedId`.

I'd personally keep `upsertedCount` in PHP for consistency with existing APIs, but I'm not going to lobby for the suggested field name in this extra text.